cpu/esp32/periph/spi.c

```diff
@@ -0,0 +1,432 @@
+/*
+ * Copyright (C) 2022 Gunar Schorcht
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_esp32
+ * @ingroup     drivers_periph_spi
+ * @{
+ *
+ * @file
+ * @brief       Low-level SPI driver implementation for ESP32 SoCs
+ *
+ * The implementation uses the ESP-IDF Low level interface in polling mode
+ * without DMA.
+ *
+ * @TODO
+ * - transaction interrupts to avoid busy waiting in polling mode
+ * - DMA transfer
+ *
+ * @author      Gunar Schorcht <gunar@schorcht.net>
+ *
+ * @}
+ */
+
+#include <assert.h>
+#include <string.h>
+
+#include "esp_common.h"
+#include "log.h"
+
+#include "cpu.h"
+#include "gpio_arch.h"
+#include "mutex.h"
+#include "periph/spi.h"
+#include "syscalls.h"
+
+#include "esp_attr.h"
+#include "esp_rom_gpio.h"
+#include "hal/spi_hal.h"
+#include "hal/spi_types.h"
+#include "soc/rtc.h"
+
+#include "esp_idf_api/periph_ctrl.h"
+
+#undef MHZ
+#include "macros/units.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+/* SPI bus descriptor structure */
+struct _spi_bus_t {
+    mutex_t lock;                    /* mutex for each SPI interface */
+    bool pins_initialized;           /* SPI pins initialized */
+    spi_host_device_t hostid;        /* SPI hostid as used by ESP-IDF */
+    const spi_signal_conn_t *periph; /* SPI peripheral descriptor */
+    spi_hal_timing_conf_t timing;    /* calculated SPI timing parameters */
+    uint32_t clk_last;               /* SPI clock speed used last time in Hz */
+    uint8_t mode_last;               /* SPI mode used last time */
+};
+
+static struct _spi_bus_t _spi[] = {
+#ifdef SPI0_CTRL
+    {
+        .pins_initialized = false,
```

Comment:
This will be the default anyway

Reply:
I know, it is the same for `.clk_last`. I have just listed them only for the sake of completeness and to make this clearer.